zookeeper/source/zk_dn_client.cc

```diff
@@ -254,6 +273,171 @@ bool ZkClientDn::poll_delete_queue() {
   return true;
 }
 
+bool ZkClientDn::poll_reconstruct_queue() {
+  return handleReconstructCmds(util::concat_path(EC_RECOVER_QUEUES,
+                                                 get_datanode_id()));
+}
+
+bool ZkClientDn::handleReconstructCmds(const std::string &path) {
+  int err;
+  std::vector<std::string> work_items;
+
+  if (!zk->get_children(path, work_items, err)) {
+    LOG(ERROR) << "Failed to get work items!";
+    return false;
+  }
+
+  if (work_items.size() > 0) {
+    LOG(INFO) << get_datanode_id()
+              << "FOUND "
+              << work_items.size()
+              << " work items";
+  }
+
+  std::vector<std::shared_ptr<ZooOp>> ops;
+  for (auto &block : work_items) {
+    auto full_work_item_path = util::concat_path(path, block);
+    std::uint64_t storage_id;
+    std::stringstream strm(block);
+    strm >> storage_id;
+    uint64_t block_group_id = ZkClientCommon::get_block_group_id(storage_id);
+    uint64_t block_idx =
+                    ZkClientCommon::get_index_within_block_group(storage_id);
+    std::vector<std::string> strg_blks = {};
+    std::string blk_grp_path =
+                    ZkClientCommon::get_block_metadata_path(block_group_id);
+    if (!zk->get_children(blk_grp_path, strg_blks, err)) {
+      LOG(ERROR) << "Could not find block group!";
+      return false;
+    }
+    std::vector<std::string> datanodes = {};
+    for (std::string strg_blk : strg_blks) {
+      zk->get_children(blk_grp_path + "/" + strg_blk, datanodes, err);
+    }
+
+    // Assume hard-coded RS(6,3), with 6 datablocks and 3 parity blocks
+    int num_data_blks = 6;
+    int num_parity_blks = 3;
+    std::vector<std::pair<uint64_t, std::string>> blk_nodes = {};
+    for (int i = 0; i < datanodes.size(); i++) {
+      bool healthy;
+      if (zk->exists(HEALTH_BACKSLASH + datanodes[i] + HEARTBEAT,
+                     healthy, err)) {
+        LOG(ERROR) << "Failed to check existence of DN";
+      }
+      uint64_t blk;
+      std::stringstream blkstrm(strg_blks[i]);
+      blkstrm >> blk;
+      if (healthy) {
+        std::pair<uint64_t, std::string> node(blk, datanodes[i]);
+        blk_nodes.push_back(node);
+//        if (blk_nodes.size() >= num_data_blks)
```

Comment:
what's this

Reply:
It's to make sure that the data node that we are getting the data/parity blocks is actually existing since we need at least 6 data nodes that are healthy!